dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/jsonpatch/BulkPatchValidators.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2004-2021, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.jsonpatch;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+import org.hisp.dhis.common.IdentifiableObject;
+import org.hisp.dhis.commons.jackson.jsonpatch.JsonPatch;
+import org.hisp.dhis.feedback.ErrorReport;
+import org.hisp.dhis.schema.Schema;
+
+/**
+ * Contains all validators needed for the json patch process
+ *
+ * @author viet@dhis2.org
+ */
+public interface BulkPatchValidators
+{
+    boolean validateJsonPatch( JsonPatch jsonPatch, Consumer<List<ErrorReport>> errorReportConsumer );
+
+    boolean validateSchema( Schema schema, Consumer<List<ErrorReport>> errorReportConsumer );
+
+    boolean validatePatchEntity( Schema schema, JsonPatch jsonPatch, String id,
+        IdentifiableObject identifiableObject, Consumer<List<ErrorReport>> errorReportConsumer );
+}
```

Comment:
I see that you changed from returning reports to consumers but I was expecting the other change we talked about changing the validator itself to something like:
```java
interface BulkPatchValidator {
  boolean validate(Schema schema, JsonPatch jsonPatch, Consumer<ErrorReport> addErrors);
}
``` 
As I indicate in the above when using a consumer it makes little sense to use `Consumer<List<T>>` since we changed to consumer to get away from the need to build intermediate objects and collection. Errors can just be added as you go. If there are many you pass them to the consumer right where they are created.

I think if you need the second level validation with the object itself and you still think this should not be done before the update in general is deemed valid it would be OK to have a validator interface with two methods.
```java
interface BulkPatchValidator {
  default boolean validate(Schema schema, JsonPatch jsonPatch, Consumer<ErrorReport> addErrors) {
    return true;
  }
  default boolean validate(Schema schema, JsonPatch jsonPatch, String id, IdentifiableObject identifiableObject, Consumer<ErrorReport> addErrors) {
    return true;
  }
}
``` 
By making them default methods one can override one of them or both when implementing a specific validator.

Then with the above type of validator interface there is no need for a `BulkPatchValidators` class.
The parameters can simply hold a `List<BulkPatchValidator>`. That was the important point so that further additions of validator do not cause changes in further classes. Instead one just uses the existing code passing other data (here validator) to it. Hope that makes it more clear.

Reply:
fixed now